pkg/controller/queuejob/queuejob_controller_ex.go

```diff
@@ -1313,134 +1298,131 @@ func (qjm *XController) ScheduleNext() {
 		}
 	}
 }
 
 // Update AppWrappers in etcd
 // todo: This is a current workaround for duplicate message bug.
-func (cc *XController) updateEtcd(qj *arbv1.AppWrapper, at string) error {
-	//apiCacheAWJob, e := cc.queueJobLister.AppWrappers(qj.Namespace).Get(qj.Name)
-	//
-	//if (e != nil) {
-	//	klog.Errorf("[updateEtcd] Failed to update status of AppWrapper %s, namespace: %s at %s err=%v",
-	//		apiCacheAWJob.Name, apiCacheAWJob.Namespace, at, e)
-	//	return e
-	//}
-
-	//TODO: Remove next line
-	var apiCacheAWJob *arbv1.AppWrapper
-	//TODO: Remove next line
-	apiCacheAWJob = qj
-	apiCacheAWJob.Status.Sender = "before " + at // set Sender string to indicate code location
-	apiCacheAWJob.Status.Local = false           // for Informer FilterFunc to pickup
-	if _, err := cc.arbclients.ArbV1().AppWrappers(apiCacheAWJob.Namespace).Update(apiCacheAWJob); err != nil {
-		klog.Errorf("[updateEtcd] Failed to update status of AppWrapper %s, namespace: %s at %s err=%v",
-			apiCacheAWJob.Name, apiCacheAWJob.Namespace, at, err)
-		return err
-		//	} else {  // qjj should be the same as qj except with newer ResourceVersion
-		//		qj.ResourceVersion = qjj.ResourceVersion  // update new ResourceVersion from etcd
+func (cc *XController) updateEtcd(currentAppwrapper *arbv1.AppWrapper, caller string) (*arbv1.AppWrapper, error) {
+	klog.V(4).Infof("[updateEtcd] trying to update '%s/%s' called by '%s'", currentAppwrapper.Namespace, currentAppwrapper.Name, caller)
+	currentAppwrapper.Status.Sender = "before " + caller // set Sender string to indicate code location
+	currentAppwrapper.Status.Local = false               // for Informer FilterFunc to pickup
+	updatedAppwrapper, err := cc.arbclients.ArbV1().AppWrappers(currentAppwrapper.Namespace).Update(currentAppwrapper)
+	if err != nil {
+		if apiErrors.IsNotFound(err) {
+			klog.Warningf("[updateEtcd] app wrapper '%s/%s' not found when called by '%s'", currentAppwrapper.Namespace, currentAppwrapper.Name, caller)
+		} else if apiErrors.IsConflict(err) {
+			klog.Warningf("[updateEtcd] app wrapper '%s/%s' update version conflict called by '%s'", currentAppwrapper.Namespace, currentAppwrapper.Name, caller)
+		}
+		klog.Errorf("[updateEtcd] Failed to update status of AppWrapper '%s/%s' called by '%s', err=%v", currentAppwrapper.Namespace, currentAppwrapper.Name, caller, err)
+		return nil, err
 	}
-	klog.V(10).Infof("[updateEtcd] AppWrapperUpdate success %s at %s &qj=%p qj=%+v",
-		apiCacheAWJob.Name, at, apiCacheAWJob, apiCacheAWJob)
-	//qj.Status.Local  = true           // for Informer FilterFunc to ignore duplicate
-	//qj.Status.Sender = "after  "+ at  // set Sender string to indicate code location
-	return nil
+	if larger(currentAppwrapper.ResourceVersion, updatedAppwrapper.ResourceVersion) {
+		klog.Warningf("[updateEtcd] updated app wrapper '%s/%s' called by '%s' has version %s", currentAppwrapper.Namespace, currentAppwrapper.Name, caller, updatedAppwrapper.ResourceVersion)
```

Comment:
I think we want to make warn log message simpler, I think we want to say that the current version of AW is greater than the working version of AW? also that larger function does a string compare rather than int compare, not sure why?